bgplgd.8
.\" $OpenBSD$
.\"
.\" Copyright (c) 2021 Claudio Jeker <claudio@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt BGPLGD 8
.Os
.Sh NAME
.Nm bgplgd
.Nd a bgpctl FastCGI server
.Sh SYNOPSIS
.Nm
.Op Fl d
.Op Fl p Ar path
.Op Fl S Ar socket
.Op Fl s Ar socket
.Op Fl U Ar user
.Op Fl u Ar user
.Sh DESCRIPTION
.Nm
is a server which implements the FastCGI Protocol to execute
.Xr bgpctl 8
commands.
.Nm
is a simple server that implements a simple web API to query
.Xr bgpd 8 .
.Pp
.Nm
opens a socket at
.Pa /var/www/run/bgplgd.sock ,
owned by www:www,
with permissions 0660.
It will then
and drop privileges to user
.Qq www ,
.Xr unveil 2
the
.Xr bgpctl 8
binary
and restrict itself with
.Xr pledge 2 .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Do not daemonize.
If this option is specified,
.Nm
will run in the foreground and log to stderr.
.It Fl p Ar path
Use
.Ar path
instead of
.Xr bgpctl 8
to query
.Xr bgpd 8 .
.It Fl S Ar socket
Use
.Ar socket
instead of the default
.Pa /var/run/bgpd.rsock
to communicate with
.Xr bgpd 8 .
.It Fl s Ar socket
Create and bind to alternative local socket at
.Ar socket .
.It Fl U Ar user
Change the owner of
.Pa /var/www/run/bgplgd.sock
to
.Ar user
and its primary group instead of the default www:www.
.It Fl u Ar user
Drop privileges to
.Ar user
instead of default user www.
.El
.Pp
.Nm
provides the following API endpoints.
Unless further specified the endpoints do not take any parameters:
.Bl -tag -width Ds
.It Pa /interfaces
Show the interface states.
.It Pa /memory
Show RIB memory statistics.
.It Pa /neighbors
Show detailed neighbors information.
The output can be limited with the following parameters:
.Bl -tag -width "neighbor=peer" -compact
.Pp
.It Cm neighbor Ns = Ns Ar peer
Show infromation for a specific neighbor.
.Ar peer
may be the neighbor's address or description.
.It Cm group Ns = Ns Ar name
Show only entries from the specified peer group.
.El
.It Pa /nexthops
Show the list of BGP nexthops and the result of their validity check.
.It Pa /rib
Show routes from the bgpd(8) Routing Information Base.
The following parameters can be used to filter the output:
.Bl -tag -width "neighbor=peer" -compact
.Pp
.It Cm neighbor Ns = Ns Ar peer
Show infromation for a specific neighbor.
.Ar peer
may be the neighbor's address or description.
.It Cm group Ns = Ns Ar name
Show only entries from the specified peer group.
.It Cm as Ns = Ns Ar number
Show only entries with the specified source AS number.
.It Cm community Ns = Ns Ar string
.It Cm ext-community Ns = Ns Ar string
.It Cm large-community Ns = Ns Ar string
Show only entries that match the specified community.
.It Xo
.Ic af Ns = Ns
.Pq Ic ipv4 | ipv6 | vpnv4 | vpnv6
.Xc
Show only entries that match the specified address family.
.It Cm rib Ns = Ns Ar name
Show only entries from the RIB with name
.Ar name .
.It Xo
.Ic ovs Ns = Ns
.Pq Ic valid | not-found | invalid
.Xc
Show only prefixes that match the specified Origin Validation State.
.It Cm best Ns = Ns 1
Show only selected routes.
.It Cm error Ns = Ns 1
Show only prefixes which are marked invalid and were treated as withdrawn.
.It Cm prefix Ns = Ns Ar addr
Show only entries that match prefix either as the best matching route or
show the entry for this CIDR prefix.
.It Cm all Ns = Ns 1
Show all entries in the specified prefix range.
.It Cm or-shorter Ns = Ns 1
Show all entries covering and including the specified prefix.
.El
.It Pa /rtr
Show a list of all RTR sessions.
.It Pa /sets
Show a list summarizing all roa-set, as-set, prefix-set, and origin-set tables.
.It Pa /summary
Show a list of all neighbors, including information about the session state
and message counters.
.El
.Sh EXAMPLES
An example setup in
.Xr httpd 8
is:
.Bd -literal -offset indent
        location "/bgplgd/*" {
                fastcgi socket "/run/bgplgd.sock"
                request strip 1
        }
.Ed
.Sh SEE ALSO
.Xr bgpctl 8 ,
.Xr bgpd 8 ,
.Xr httpd 8
.Sh HISTORY
The
.Nm
server first appeared in
.Ox 7.0 .
.Sh AUTHORS
.An Claudio Jeker Aq Mt claudio@openbsd.org


bgplgd.c
/*	$OpenBSD$ */
/*
 * Copyright (c) 2020 Claudio Jeker <claudio@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "bgplgd.h"
#include "slowcgi.h"
#include "http.h"

#define NCMDARGS	4

const struct cmd {
	const char	*path;
	char		*args[NCMDARGS];
	unsigned int	qs_mask;
	int		barenbr;
} cmds[] = {
	{ "/interfaces", { "show", "interfaces", NULL }, 0 },
	{ "/memory", { "show", "rib", "memory", NULL }, 0 },
	{ "/neighbors", { "show", "neighbor", NULL }, QS_MASK_NEIGHBOR, 1 },
	{ "/nexthops", { "show", "nexthop", NULL }, 0 },
	{ "/rib", { "show", "rib", "detail", NULL }, QS_MASK_RIB },
	{ "/rtr", { "show", "rtr", NULL }, 0 },
	{ "/sets", { "show", "sets", NULL }, 0 },
	{ "/summary", { "show", NULL }, 0 },
	{ NULL }
};

static const char *
http_error(int *res)
{
	const struct http_error errors[] = HTTP_ERRORS;
	size_t i;

	for (i = 0; errors[i].error_code != 0; i++)
		if (errors[i].error_code == *res)
			return errors[i].error_name;

	/* unknown error - change to 500 */
	lwarnx("unknown http error %d", *res);
	*res = 500;
	return "Internal Server Error";
}

static void
error_response(int res)
{
	const char *type = "text/html";
	const char *errstr = http_error(&res);

	lwarnx("HTTP status %d: %s", res, errstr);

	printf(
	    "Content-Type: %s\n"
	    "Status: %d\n"
	    "Cache-Control: no-cache\n"
	    "\n"
	    "<!DOCTYPE html>\n"
	    "<html>\n"
	    " <head>\n"
	    "  <meta http-equiv=\"Content-Type\" "
	    "content=\"%s; charset=utf-8\"/>\n"
	    "  <title>%d %s</title>\n"
	    " </head>\n"
	    " <body>\n"
	    "  <h1>%d %s</h1>\n"
	    "  <hr>\n"
	    "  <address>OpenBSD bgplgd</address>\n"
	    " </body>\n"
	    "</html>\n",
	    type, res, type, res, errstr, res, errstr);

	exit(0);
}

static int
command_from_path(const char *path, struct lg_ctx *ctx)
{
	size_t i;

	for (i = 0; cmds[i].path != NULL; i++) {
		if (strcmp(cmds[i].path, path) == 0) {
			ctx->command = &cmds[i];
			ctx->qs_mask = cmds[i].qs_mask;
			return 0;
		}
	}
	return 404;
}

static int
prep_request(struct lg_ctx *ctx, const char *meth, const char *path,
    const char *qs)
{
	if (meth == NULL || path == NULL)
		return 500;
	if (strcmp(meth, "GET") != 0)
		return 405;

	if (command_from_path(path, ctx) != 0)
		return 404;

	if (parse_querystring(qs, ctx) != 0)
		return 400;

	return 0;
}


/*
 * Entry point from the FastCGI handler.
 * This runs as an own process and can use STDOUT and STDERR.
 */
void
call(const char *method, const char *pathinfo, const char *querystring)
{
	struct lg_ctx ctx;
	char *argv[64];
	size_t i, argc = 0;
	int res;

	memset(&ctx, 0, sizeof(ctx));
	if ((res = prep_request(&ctx, method, pathinfo, querystring)) != 0)
		error_response(res);

	argv[argc++] = bgpctlpath;
	argv[argc++] = "-j";
	argv[argc++] = "-s";
	argv[argc++] = bgpctlsock;

	for (i = 0; ctx.command->args[i] != NULL; i++)
		argv[argc++] = ctx.command->args[i];

	argc = qs_argv(argv, argc, sizeof(argv) / sizeof(argv[0]), &ctx,
	    ctx.command->barenbr);

	argv[argc++] = NULL;

	for (i = 0; argv[i] != NULL; i++)
		ldebug("argv[%zu], %s", i, argv[i]);


	signal(SIGPIPE, SIG_DFL);

	/* Write server header first */
	printf("Content-type: application/json\r\n\r\n");
	fflush(stdout);

	execvp(bgpctlpath, argv);

	lerr(1, "failed to execute bgpctl");
}
